File: CVS_ObjectDetectorSample_ObjC/CVS_ObjectDetectorSample/BoundingBoxLayer.m
// Copyright (c) Microsoft Corporation.  All rights reserved.

#import "BoundingBoxLayer.h"
#import "BoundingBoxLabelLayer.h"

@interface BoundingBoxLayer ()

@property (strong, nonatomic) BoundingBoxLabelLayer *labelLayer;

@end


@implementation BoundingBoxLayer

- (instancetype)init
{
    self = [super init];
    if (self)
    {
        self.backgroundColor = [[UIColor clearColor] CGColor];
        self.borderColor = [[UIColor whiteColor] CGColor];
        self.borderWidth = 4.0f;
        
        self.labelLayer = [[BoundingBoxLabelLayer alloc] init];
        self.labelLayer.backgroundColor = [[UIColor whiteColor] CGColor];
        self.labelLayer.foregroundColor = [[UIColor blackColor] CGColor];
        self.labelLayer.contentsScale = [[UIScreen mainScreen] scale];
        self.labelLayer.font = (__bridge CFStringRef)[[UIFont systemFontOfSize:12] fontName];
        self.labelLayer.fontSize = 14.0f;
        self.labelLayer.alignmentMode = kCAAlignmentLeft;
        self.labelLayer.padding = CGSizeMake(16.0f, 8.0f);

        if (@available(iOS 11.0, *)) {
            self.cornerRadius = 6.0f;
            self.maskedCorners = kCALayerMaxXMinYCorner | kCALayerMinXMaxYCorner | kCALayerMaxXMaxYCorner;

            self.labelLayer.cornerRadius = 6.0f;
            self.labelLayer.maskedCorners = kCALayerMinXMinYCorner | kCALayerMaxXMinYCorner;
        }

        [self addSublayer:self.labelLayer];
    }
    return self;
}

- (void)setLabel:(NSString *)label
{
    _label = label;

    self.labelLayer.string = label;
    [self setNeedsLayout];
}

- (void)setColor:(UIColor *)color
{
    _color = color;
    
    //
    
    self.borderColor = [color CGColor];
    self.labelLayer.backgroundColor = [color CGColor];
}

- (void)layoutSublayers
{
    CGSize labelSize = self.labelLayer.preferredFrameSize;
    self.labelLayer.frame = CGRectMake(0.0f, - labelSize.height + self.borderWidth, labelSize.width, labelSize.height);
}

@end

File: CVS_ObjectDetectorSample_Swift/CVS_ObjectDetectorSample/BoundingBoxLayer.swift
// Copyright (c) Microsoft Corporation.  All rights reserved.

import UIKit

class BoundingBoxLayer: CALayer {

    var label: String? {
        didSet {
            labelLayer.string = label
            setNeedsLayout()
        }
    }
    
    var color: UIColor = UIColor.white {
        didSet {
            borderColor = color.cgColor
            labelLayer.backgroundColor = color.cgColor
        }
    }
    
    private var labelLayer = BoundingBoxLabelLayer()
    
    override init() {
        super.init()
        
        backgroundColor = UIColor.clear.cgColor
        borderColor = color.cgColor
        borderWidth = 4.0

        labelLayer.backgroundColor = color.cgColor
        labelLayer.foregroundColor = UIColor.black.cgColor
        labelLayer.contentsScale = UIScreen.main.scale
        labelLayer.font = UIFont.systemFont(ofSize: UIFont.systemFontSize).fontName as CFTypeRef
        labelLayer.fontSize = 14.0
        labelLayer.alignmentMode = CATextLayerAlignmentMode.left
        labelLayer.padding = CGSize.init(width: 16.0, height: 8.0)

        if #available(iOS 11.0, *) {
            cornerRadius = 6.0
            maskedCorners = [.layerMaxXMinYCorner, .layerMinXMaxYCorner, .layerMaxXMaxYCorner ]

            labelLayer.cornerRadius = 6.0
            labelLayer.maskedCorners = [.layerMinXMinYCorner, .layerMaxXMinYCorner]
        }

        addSublayer(labelLayer)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func layoutSublayers() {

        let labelSize = labelLayer.preferredFrameSize();
        labelLayer.frame = CGRect.init(
            x: 0.0,
            y: -labelSize.height + self.borderWidth,
            width: labelSize.width,
            height: labelSize.height);
    }

}
